### PR TITLE
Update to Android v5

### PR DESCRIPTION
**Pull request by beylmk**
Android equivalent of https://github.com/RevenueCat/purchases-hybrid-common/pull/113

This was branched off `update-to-rc2` and then main was merged in... 

Adds:
* setLogHandler
* dangerousSettings option in configuration

Also:
* pulls fixes from https://github.com/RevenueCat/purchases-hybrid-common/pull/106 into the renamed StoreProductMappers file
* defaults the purchases-android dependency to the `latestDependencies` flavor... not sure if we want to make flavors for purchases-hybrid-common, but figured that would be in a separate PR if so

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -156,3 +156,40 @@ workflows:
...
+            branches:
+              ignore: /.*/
+      - deploy-android:
+          requires:
+            - test-android
+          filters:
+            tags:
+              only: /^(\d+\.)(\d+\.)(\d+)$/
+            branches:
+              ignore: /.*/
```
Comment (aboedo) [inline .circleci/config.yml:175]: how does this differ from the `deploy-android` job?

File: .circleci/config.yml [hunk 2]
```diff
@@ -156,3 +156,40 @@ workflows:
...
+  android-deploy-amazon:
+    jobs:
+      - test-android:
+          filters:
+            branches:
+              only: amazon
+      - hold:
+          type: approval
+          requires:
+            - test-android
```
Comment (aboedo) [inline .circleci/config.yml:186]: why does this one require approval vs the others?

Comment (aboedo) [inline .circleci/config.yml:186]: why does this one require approval vs the others?

File: android/build.gradle [hunk 3]
```diff
@@ -39,6 +39,7 @@ android {
         versionName "2.0.1"
 
         consumerProguardFiles 'consumer-rules.pro'
+        missingDimensionStrategy 'dependencyVersions', 'latestDependencies', 'unityIAP'
```
Comment (aboedo) [inline android/build.gradle:42]: Could we add a comment here specifying what it does? 

I figure that this will try to choose the latest dependencies first, but if there's an incompatibility with other requirements for the same app, it'll try the flavor `unityIAP`, is that correct?

Comment (aboedo) [inline android/build.gradle:42]: Could we add a comment here specifying what it does? 

My understanding is that this will try to choose the latest dependencies first, but if there's an incompatibility with other requirements for the same app, it'll try the flavor `unityIAP`, is that correct?

File: android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
+package com.revenuecat.purchases.hybridcommon.mappers
+
+import com.revenuecat.purchases.models.StoreProduct
+
+fun StoreProduct.map(): Map<String, Any?> =
+    mapOf(
+        "identifier" to sku,
+        "description" to description,
+        "title" to title,
+        "price" to priceAmountMicros / 1000000.0,
```
Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt:10]: I think this was here before but could we have this be `1_000_000.0` for readability?

Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt:10]: I think this was here before but could we have this be `1_000_000.0` for readability?

File: android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                    "priceString" to formatUsingDeviceLocale(priceCurrencyCode, 0),
+                    "period" to freeTrialPeriod,
+                    "cycles" to 1
+                ) + periodFields
+            } ?: null
+        }
+        introductoryPrice != null -> {
+            introductoryPricePeriod!!.mapPeriod()?.let { periodFields ->
+                mapOf(
+                    "price" to introductoryPriceAmountMicros / 1000000.0,
```
Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt:37]: same here

Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/mappers/StoreProductMapper.kt:37]: same here

File: .circleci/config.yml (hunk 1 above)
Comment (vegaro) [reply .circleci/config.yml:175]: we don't need these, removing. I think they are leftovers from manually triggered releases we've been doing in the past year

File: android/build.gradle (hunk 3 above)
Comment (vegaro) [reply android/build.gradle:42]: I am not sure what this is for either @beylmk . Can you add some context?

File: .circleci/config.yml (hunk 2 above)
Comment (vegaro) [reply .circleci/config.yml]: Same as the other comment. We were triggering releases using the approval step, I don't remember why. I removed it because it's not needed

File: android/build.gradle (hunk 3 above)
Comment (beylmk) [reply android/build.gradle]: @aboedo yes, that's correct. we have to specify the flavor of purchases-android that we want to depend on, and this defaults to latestDependencies. 

we could instead make purchases-hybrid-common flavors, which in turn pull a specific purchases-android flavor. i figured that could be a follow-up ticket

Comment (beylmk) [reply android/build.gradle]: see points 2/3 in the TLDR at this link: https://kiranrao.in/blog/2020/03/31/gradle-missing-flavors/

according to that, we probably do want to make PHC flavors

Comment (vegaro) [reply android/build.gradle]: I am actually working on adding the flavors so we can improve Unity by A LOT

Comment (vegaro) [reply android/build.gradle]: I think maybe this was required because you were doing local development and pointing to your copy of purchases-android 🤔

File: android/src/main/java/com/revenuecat/purchases/hybridcommon/common.kt [hunk 6]
```diff
@@ -194,9 +194,9 @@ fun logIn(
 ) {
```
Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/common.kt:194]: we should also rename `restoreTransactions` -> `restorePurchases`, but that could be a separate PR

File: android/src/main/java/com/revenuecat/purchases/hybridcommon/common.kt [hunk 7]
```diff
@@ -322,14 +288,21 @@ fun configure(
...
-        Purchases.configure(context, apiKey, appUserID, observerMode)
-    } else {
-        Purchases.configure(context, apiKey, appUserID)
+        builder.observerMode(observerMode)
     }
+
+    Purchases.configure(builder.build())
 }
 
 fun getPaymentDiscount() : ErrorContainer {
```
Comment (aboedo) [inline android/src/main/java/com/revenuecat/purchases/hybridcommon/common.kt:308]: this should be renamed to getPromotionalOffer, but it can also be done as a separate PR

**Review by aboedo — APPROVED** [review]
Looks great! I left a couple of notes w/regards to other APIs that need a rename, but we can do that as a separate thing. 
Have you already tested this on a hybrid?

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply .circleci/config.yml]: lol I guess I had that comment for weeks and hadn't posted the PR 🤦‍♂️

**Comment by vegaro** [comment]
will open follow up PRs. I have tested this while working on updating Flutter and so far it's compiling. Some things to change on the Flutter side though 😄
